### PR TITLE
watchdog: properly finish goroutine for periodic integrity checks

### PR DESCRIPTION
**Pull request by AlgebraicWolf**
This patch properly finalises goroutines performing integrity checks. This prevents hanging of watchdog upon stopping the instance with integrity checks turned on.

### CODE REVIEW

**Comment by AlgebraicWolf** [comment]
Be not fooled by green CI, for it seems I might have broken `tt stop` without integrity checks enabled. I shall investigate and mark PR as ready once I've figured it out.

File: cli/running/watchdog.go [hunk 1]
```diff
@@ -39,6 +39,9 @@ type Watchdog struct {
 	// done channel used to inform the signal handle goroutine
 	// about termination of the Instance.
 	done chan bool
+	// doneIntegrity channel used to inform the integrity checking
+	// goroutine about termination of the instance
+	doneIntegrity chan bool
```
[the remark this answers is not in the document]
Comment (AlgebraicWolf) [reply cli/running/watchdog.go]: I've tried it, and it is more convenient indeed :) 
Besides, now `stop` works properly both with and without integrity checks. I am undrafting this PR.

File: cli/running/watchdog.go [hunk 2]
```diff
@@ -89,8 +92,8 @@ func (wd *Watchdog) Start() error {
 
 	if err = wd.preStartAction(); err != nil {
 		wd.logger.Printf(`(ERROR): Pre-start action error: %v`, err)
-		// Finish the signal handling goroutine.
-		wd.done <- true
+		// Finish the signal handling and periodic integrity checking goroutines.
+		wd.goroutineCancel()
```
Comment (oleg-jukovec) [inline cli/running/watchdog.go:96]: The key difference in the code: it does not wait for the message receiving now.

File: cli/running/watchdog.go [hunk 3]
```diff
@@ -36,9 +37,11 @@ type Watchdog struct {
...
 	// restarting the Instance.
 	restartTimeout time.Duration
-	// done channel used to inform the signal handle goroutine
-	// about termination of the Instance.
-	done chan bool
+	// goroutineCtx controls goroutines that receive signals and perform periodic
+	// integrity checks.
+	goroutineCtx context.Context
+	// goroutineCancel stops goroutines associated with ctx.
+	goroutineCancel context.CancelFunc
```
[the remark this answers is not in the document]
Comment (AlgebraicWolf) [reply cli/running/watchdog.go]: Fixed!

File: cli/running/watchdog.go [hunk 4]
```diff
@@ -78,19 +75,20 @@ func (wd *Watchdog) Start() error {
...
-		wd.startIntegrityChecks()
+		wd.startIntegrityChecks(goroutineCtx)
 	}
 
 	if err = wd.preStartAction(); err != nil {
 		wd.logger.Printf(`(ERROR): Pre-start action error: %v`, err)
-		// Finish the signal handling goroutine.
-		wd.done <- true
+		// Finish the signal handling and periodic integrity checking goroutines.
+		goroutineCancel()
```
Comment (oleg-jukovec) [inline cli/running/watchdog.go:91]: Do we need to wait goroutines here?
```suggestion
		goroutineCancel()
		wd.doneBarrier.Wait()
```

Comment (AlgebraicWolf) [reply cli/running/watchdog.go:91]: We can and it doesn't seem to hurt, although I guess it's not really necessary -- it seems to me that the only place we need to synchronise with the goroutines is when we are restarting them, and we don't do this here.

Comment (AlgebraicWolf) [reply cli/running/watchdog.go:91]: I've added the synchronization here anyway. Maybe it will save us some pain in the future.

**Review by oleg-jukovec — APPROVED** [review]

File: cli/running/watchdog.go [hunk 5]
```diff
@@ -60,14 +58,13 @@ func NewWatchdog(restartable bool, restartTimeout time.Duration, logger *ttlog.L
...
 
-	wd.done = make(chan bool, 1)
-
 	return &wd
 }
 
 // Start starts the Instance and signal handling.
 func (wd *Watchdog) Start() error {
 	var err error
+	// var goroutineCtx context.Context
```
Comment (psergee) [inline cli/running/watchdog.go:67]: Pls, remove this comment.

File: cli/running/watchdog.go [hunk 6]
```diff
@@ -78,19 +75,21 @@ func (wd *Watchdog) Start() error {
 	// get started for avoiding a race condition between tt start
 	// and tt stop. This way we avoid a situation when we receive
 	// a signal before starting a handler for it.
-	wd.startSignalHandling()
+	goroutineCtx, goroutineCancel := context.WithCancel(context.Background())
```
Comment (psergee) [inline cli/running/watchdog.go:78]: Can we be more specific in names? All contexts are related to goroutines :) We can use `watchdogCtx` and `watchdogCancel` for example here. Or something similar that will give us a clue of context and cancel func purpose. Compare: `watchdogCancel()` and `goroutineCancel()`.

**Review by psergee — APPROVED** [review]
Thank you for the patch! Minor comments left.

File: cli/running/watchdog.go (hunk 5 above)
Comment (AlgebraicWolf) [reply cli/running/watchdog.go]: Oops! Fixed.

File: cli/running/watchdog.go (hunk 6 above)
Comment (AlgebraicWolf) [reply cli/running/watchdog.go]: Makes sense. I've changed it to `watchdogCtx` and `watchdogCancel`

**Comment by AlgebraicWolf** [comment]
Looks like something is wrong with the commit used in  `test_install_tarantool_commit` integration test.

**Comment by psergee** [comment]
> Looks like something is wrong with the commit used in `test_install_tarantool_commit` integration test.

Fixed. Pls, rebase.

**Pull request merged**